spring-cloud-task-core/src/main/java/org/springframework/cloud/task/listener/TaskLifecycleListener.java

```diff
@@ -0,0 +1,162 @@
+ *
+ * <ul>
+ *     <li>{@link ContextRefreshedEvent} - Used to identify the start of a task.  A task
+ *     is expected to contain a single application context.</li>
+ *     <li>{@link ContextClosedEvent} - Used to identify the successful end of a task.</li>
+ *     <li>{@link ApplicationFailedEvent} - Used to identify the failure of a task.</li>
+ * </ul>
+ *
+ * <b>NOTE:</b> Multiple contexts (including parent/child relationships) will result in
+ * only the first context refresh that contains this instance being recorded.
+ *
+ * @author Michael Minella
+ */
+public class TaskLifecycleListener implements ApplicationListener<ApplicationEvent>,
+		ApplicationContextAware {
+
+	private final static Logger logger = LoggerFactory.getLogger(TaskLifecycleListener.class);
+
+	private final TaskRepository taskRepository;
+
+	private TaskExecution taskExecution;
+
+	private boolean started = false;
+
+	private ApplicationContext applicationContext;
+
+	/**
+	 * @param taskRepository The repository to record executions in.
+	 */
+	public TaskLifecycleListener(TaskRepository taskRepository) {
+		this.taskRepository = taskRepository;
+	}
+
+	/**
+	 * @param applicationEvent The application being listened for.
+	 */
+	@Override
+	public void onApplicationEvent(ApplicationEvent applicationEvent) {
+		if(applicationEvent instanceof ContextRefreshedEvent) {
+			doTaskStart();
+			started = true;
+		}
+		else if(applicationEvent instanceof ContextClosedEvent) {
+			doTaskEnd();
+		}
+		else if(applicationEvent instanceof ApplicationFailedEvent) {
+			doTaskFailed(((ApplicationFailedEvent) applicationEvent).getException());
+		}
+	}
+
+	private void doTaskFailed(Throwable exception) {
+
+		if(started) {
+			this.taskExecution.setEndTime(new Date());
+			//TODO: get exit code from new event thrown by Boot.
+			this.taskExecution.setExitCode(1);
+			this.taskExecution.setExitMessage(stackTraceToString(exception));
+
+			taskRepository.update(taskExecution);
+		}
+		else  {
+			logger.warn("An event to fail a task has been received for a task that has " +
+					"not yet started.");
+		}
+	}
+
+	private String stackTraceToString(Throwable exception) {
+		StringWriter writer = new StringWriter();
+		PrintWriter printWriter = new PrintWriter(writer);
+
+		exception.printStackTrace(printWriter);
+
+		return writer.toString();
+	}
+
+	private void doTaskEnd() {
+		if(started) {
+			this.taskExecution.setEndTime(new Date());
+			//TODO: get exit code from new event thrown by Boot.
+			this.taskExecution.setExitCode(0);
+
+			taskRepository.update(taskExecution);
+		}
+		else {
+			logger.warn("An event to end a task has been received for a task that has " +
+					"not yet started.");
+		}
+	}
+
+	private void doTaskStart() {
+
+		if(!started) {
+
+			String executionId = UUID.randomUUID().toString();
+			this.taskExecution = new TaskExecution();
+
+			this.taskExecution.setTaskName(applicationContext.getId());
```

Comment:
As discussed, this is the expected way to obtain the fully processed value per https://github.com/spring-projects/spring-boot/blob/master/spring-boot/src/main/java/org/springframework/boot/context/ContextIdApplicationContextInitializer.java